File: lee/src/main/java/org/example/lee/题目/链表/b_随机链表的复制.java
package org.example.lee.题目.链表;

import java.util.HashMap;
import org.example.lee.model.重名.Node;

public class b_随机链表的复制 {

  /**
   * copy random list 轻松拿下
   *
   * @param head head
   * @return {@link Node}
   */
  public Node copyRandomList(Node head) {
    if (head == null) {
      return null;
    }
    HashMap<Node, Node> map2 = new HashMap<>();
    Node node = new Node(head.val);
    Node save = node, save1 = head;
    map2.put(head, node);
    head = head.next;
    while (head != null) {
      node = new Node(head.val);
      map2.put(head, node);
      head = head.next;
    }
    while (save1 != null) {
      Node node2 = map2.get(save1);
      node2.next = map2.get(save1.next);
      node2.random = map2.get(save1.random);
      save1 = save1.next;
    }
    return save;
  }
}

File: lee/src/main/java/org/example/lee/题目/链表/b_随机链表的复制.go
package 链表

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	m := make(map[*Node]*Node)
	save := head
	for head != nil {
		m[head] = &Node{
			Val:    head.Val,
			Next:   nil,
			Random: nil,
		}
		head = head.Next
	}
	head = save
	var newHead *Node
	for head != nil {
		newHead = m[head]
		newHead.Next = m[head.Next]
		newHead.Random = m[head.Random]
		head = head.Next
	}
	return m[save]
}
